[SILDiagnostics] Add static enforcement of Law of Exclusivity

devincoughlin opened the pull request:
Add a diagnostic pass that emits errors when a violation of the "Law of
Exclusivity" is detected at compile time. The Law of Exclusivity requires
that the access duration of any access to an address not overlap
with an access to the same address unless both accesses are reads.

This pass relies on 'begin_access' and 'end_access' SIL instruction
markers inserted by SILGen to determine when an access to an address begins and
ends. It models the in-progress accesses with a stack of 'begin_access'
instructions.

gottesmm commented:
Where does this fit into the pipeline? Will it require any changes for me to move the ownership model eliminator past it?

atrick commented:
Assuming ownership elimination will runs immediately after SILGen, Devin should ignore it. (That means no `copy_value` instructions, unless we add special support for them).

devincoughlin commented:
I discussed this with Andy and Michael offline. Michael pointed out that we can get into situations where using the project_box instruction as the identity of a boxed local will result in false negatives. For example, it is possible project from a box and then later copy the box with copy_aclue and project from the copy. Then the two 'project_box'es represent the same location but are different instructions.

To handle this we'll need to track back from the begin_access'd operand to find the underlying alloc_box. This same kind of tracking would be needed for a best effort static enforcement of dynamic accesses. I'll update the patch to do this for the locals case and lay the groundwork globals, etc.

atrick reviewed (commented):
This looks pretty good. Here are the major points:

1. We talked about how we need to track storage locations, not
   addresses since we could have multipe SSA instances of the same
   address value. i.e. multiple projections from the same box. It
   might be expensive for the SIL verifier to check that multiple
   projections of the same location don't occur within an
   access. Whereas simply supporting multiple projections might be easy.

2. I thought stack discipline would only apply to accesses for a given
   storage location. I see no reason that independent storage
   locations would need to follow stack discipline across their
   accesses. If we really think this is a desirable structural SIL
   property, I'm sure we could support it. If it becomes too big of an
   annoyance in the optimizer we could loosen it later by adding more
   support to the checker. But is it really desirable?

3. Related to #1, if you track access per location, I think you can
   just use counters rather than a stack (# reads, # writes)

4. I'm questioning the wisdom of assuming that we have very few
   "in-flight" accesses because each time we inline something with in
   'inout' arg we get another level of nesting. It probably won't be
   to bad, but with per location counters it would be a non-issue.

5. Iterative dataflow hinging on accessesChanged doesn't seem like the
   right approach (it's hard for me to see how accessesChanged can be
   both sound and efficient). We don't need iterative data flow since
   the access state needs to be the same on all edges in or out of a
   block. If you process the blocks once following RPO the you can
   simply compare (processed) incoming edges before visiting a block
   and assert. You can make assertions about backedges easily after
   processing a block. Processing the blocks in RPO will give you
   predictable diagnostics (across compiler revisions assuming branch
   targets aren't changed), and I don't think there's anything special
   about the Function's incidental block order once we start running
   SIL passes.

devincoughlin commented:
Thanks Andy!

> I see no reason that independent storage
> locations would need to follow stack discipline across their
> accesses.

I will loosen this requirement in the checker, but I think the stack discipline is a reasonable property to enforce in the verifier.  This is because In general we can't tell whether two SIL address values represent independent storage locations. SILGen-generated SIL should have this follow the stack discipline (since accesses are lexically scoped). If we maintain the discipline then we can have much higher quality dynamic diagnostics (for example, on a failed attempt to gain an exclusive write we could print the PC/line number where the outermost conflicting read began). In debug builds, at least, this seems like it would be very valuable.

> I'm questioning the wisdom of assuming that we have very few
> "in-flight" accesses because each time we inline something with in
> 'inout' arg we get another level of nesting. It probably won't be
> to bad, but with per location counters it would be a non-issue.

What data structure would you suggest for per-location counters? A hash from locations (ex., an alloc_box instruction) to a pair of counts? We're running this before inlining, so I'm not too worried about large numbers of in-flight accesses -- but I'm happy to change it to avoid unexpected surprises.

> Iterative dataflow hinging on accessesChanged doesn't seem like the
> right approach (it's hard for me to see how accessesChanged can be
> both sound and efficient). 

Thanks, I'll change to an RPO traversal.

atrick commented:
> What data structure would you suggest for per-location counters? A hash from locations (ex., an alloc_box instruction) to a pair of counts? We're running this before inlining, so I'm not too worried about large numbers of in-flight accesses -- but I'm happy to change it to avoid unexpected surprises.

Yes. IIRC you already hash the instruction to see if it's been diagnosed, although that won't be necessary if you do an RPO traversal instead of iterating.

atrick commented:
> I will loosen this requirement in the checker, but I think the stack discipline is a reasonable property to enforce in the verifier. This is because In general we can't tell whether two SIL address values represent independent storage locations. SILGen-generated SIL should have this follow the stack discipline (since accesses are lexically scoped). If we maintain the discipline then we can have much higher quality dynamic diagnostics (for example, on a failed attempt to gain an exclusive write we could print the PC/line number where the outermost conflicting read began). In debug builds, at least, this seems like it would be very valuable.

Fair points. Better dynamic diagnostics is compelling.

devincoughlin commented:
I've updated this to address Andy and Michael's feedback:

- It now models the in-progress accesses with a mapping from accessed storage to counts of read and write-like accesses to that storage. To preserve diagnostic fidelity it also tracks the instruction that began the first in-progress access for the storage.
- It looks through 'project_box' and 'copy_value' instructions to identify the root 'alloc_box' being accessed.
- It uses a single RPO traversal rather than iterative data flow.
- The tests no longer assume ownership unqualified sil.

atrick commented:
Nice!

It's much more clear now how the data flow progresses through the CFG.

As we discussed, RPO isn't strictly necessary. Since you only need
*any* path to reach a block before it is processed. But since the cost
of DFS is insignificant relative to a pass that actually visits all
instructions, it's not really a concern.

You're completely leaving it up to the SIL verifier now to ensure the
accesses are well-formed w.r.t control flow. (I somewhat misunderstood
that in my first review). I think that's a good separation of
concerns. But any assumptions made by this pass should probably be
clearly commented, even if they are checked by the verification pass.

devincoughlin committed:
[SILDiagnostics] Add static enforcement of Law of Exclusivity

Add a diagnostic pass that emits errors when a violation of the "Law of
Exclusivity" is detected at compile time. The Law of Exclusivity requires
that the access duration of any access to an address not overlap
with an access to the same address unless both accesses are reads.

This pass relies on 'begin_access' and 'end_access' SIL instruction
markers inserted by SILGen to determine when an access to an address begins and
ends. It models the in-progress accesses with a map from storage locations to
the counts of read and write-like accesses in progress for that location.

devincoughlin commented:
Updated to include a comment describing the key SIL invariants the pass relies upon.

devincoughlin commented:
@swift-ci Please smoke test and merge

devincoughlin commented:
@swift-ci Please smoke test

gottesmm commented on lib/SILOptimizer/Mandatory/DiagnoseStaticExclusivity.cpp:
```diff
@@ -0,0 +1,349 @@
+
+  BlockOutAccesses[Fn.getEntryBlock()] = StorageMap();
+
+  llvm::ReversePostOrderTraversal<SILFunction *> RPOT(&Fn);
```
One comment. You shouldn't use this. We have an analysis for computing a reverse post order. You should use that as well. This will ensure that if someone creates an RPO before you or after you reuse of the RPO can happen.

gottesmm commented on lib/SILOptimizer/Mandatory/DiagnoseStaticExclusivity.cpp:
```diff
@@ -0,0 +1,349 @@
+           NewAccess->getAccessKind() == SILAccessKind::Read));
+
+  ExclusiveOrShared_t NewRequires;
+  if (NewAccess->getAccessKind() == SILAccessKind::Read) {
```
Can you refactor this out into a helper function.

gottesmm commented on lib/SILOptimizer/Mandatory/DiagnoseStaticExclusivity.cpp:
```diff
@@ -0,0 +1,349 @@
+  }
+}
+
+using StorageMap = llvm::SmallDenseMap<AccessedStorage, AccessInfo, 4>;
```
Can you put this in an anonymous namespace.

gottesmm commented on lib/SILOptimizer/Mandatory/DiagnoseStaticExclusivity.cpp:
```diff
@@ -0,0 +1,349 @@
+    // location.
+    StorageMap &Accesses = *BBState;
+
+    for (auto I = BB->begin(), E = BB->end(); I != E; ++I) {
```
Can you use a foreach loop here?

gottesmm commented on lib/SILOptimizer/Mandatory/DiagnoseStaticExclusivity.cpp:
```diff
@@ -0,0 +1,349 @@
+        }
+
+        Info.beginAccess(BAI);
+      } else if (auto *EAI = dyn_cast<EndAccessInst>(I)) {
```
Rather than doing an else if chain here, can you do:

```
if (...) {
  continue
}

if (...) {
  continue
}
```

gottesmm commented on lib/SILOptimizer/Mandatory/DiagnoseStaticExclusivity.cpp:
```diff
@@ -0,0 +1,349 @@
+    if (Kind == SILAccessKind::Read) {
+      // A read conflicts with any non-read accesses.
+      return NonReads > 0;
+    } else {
```
This violates llvm style. The else case is not needed.

gottesmm reviewed (commented):
A few more style issues, reviewed post commit. The earlier comment about using the analysis is a real issue though. We should try and reuse RPOT if we can.

devincoughlin commented:
Thanks Michael!

The RPO issue is addressed in https://github.com/apple/swift/pull/8670 and the style issues in https://github.com/apple/swift/pull/8662
